tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+        return {"msg": "PO No URL Match"}
+
+
+@inlineCallbacks
+def test_url_only_no_match() -> None:
+    """Same case as with ``test_url_only_match()`` but the URL specified in the
+    ``@handle_urls`` annotation doesn't match the request/response URL that the
+    spider is crawling.
+
+    However, it should still work since we're forcing to use ``UrlNoMatchPage``
+    specifically as the page object input.
+    """
+    item, deps = yield crawl_item_and_deps(UrlNoMatchPage)
+    assert item == {"msg": "PO No URL Match"}
+    assert_deps(deps, {"page": UrlNoMatchPage})
+
+
+class NoRulePage(ItemPage):
+    async def to_item(self) -> dict:
+        return {"msg": "NO Rule"}
+
+
+class NoRuleWebPage(WebPage):
+    async def to_item(self) -> dict:
+        return {"msg": "NO Rule Web"}
+
+
+@inlineCallbacks
+def test_no_rule_declaration() -> None:
+    """A more extreme case of ``test_url_only_no_match()`` where the page object
+    doesn't have any rule declaration at all.
+
+    But it should still work since we're enforcing the dependency.
+    """
+    item, deps = yield crawl_item_and_deps(NoRulePage)
+    assert item == {"msg": "NO Rule"}
+    assert_deps(deps, {"page": NoRulePage})
+
+    item, deps = yield crawl_item_and_deps(NoRuleWebPage)
+    assert item == {"msg": "NO Rule Web"}
+    assert_deps(deps, {"page": NoRuleWebPage})
+
+
+class OverriddenPage(WebPage):
+    async def to_item(self) -> dict:
+        return {"msg": "PO that will be replaced"}
+
+
+@handle_urls(URL, instead_of=OverriddenPage)
+class ReplacementPage(WebPage):
+    async def to_item(self) -> dict:
+        return {"msg": "PO replacement"}
+
+
+@inlineCallbacks
+def test_basic_overrides() -> None:
+    """Basic overrides use case.
+
+    If a page object is asked for and it's available in a rule's ``instead_of``
+    parameter, it would be replaced by the page object inside the rule's ``use``
+    parameter.
+    """
+    item, deps = yield crawl_item_and_deps(OverriddenPage)
+    assert item == {"msg": "PO replacement"}
+    assert_deps(deps, {"page": ReplacementPage})
+
+    # Calling the replacement should also still work
+    item, deps = yield crawl_item_and_deps(ReplacementPage)
+    assert item == {"msg": "PO replacement"}
+    assert_deps(deps, {"page": ReplacementPage})
+
+
+class LeftPage(WebPage):
+    async def to_item(self) -> dict:
+        return {"msg": "left page"}
+
+
+@handle_urls(URL, instead_of=LeftPage)
+class RightPage(WebPage):
+    async def to_item(self) -> dict:
+        return {"msg": "right page"}
+
+
+# We define it here since it errors out if made as a decorator
+# since RightPage doesn't exist yet.
+handle_urls(URL, instead_of=RightPage)(LeftPage)
+
+
+@inlineCallbacks
+def test_mutual_overrides() -> None:
+    """Two page objects that override each other should not present any problems.
+
+    In practice, this isn't useful at all.
+
+    TODO: We could present a warning to the user if this is detected.
```

Comment:
Raising an error is also a good option, might be even better. I don't see a use case for this; it seems this can get confusing.

Reply:
Though no strong opinion on that.

Reply:
Yeah, leaving it as-is for now until users do raise use cases for them.